src/hotspot/cpu/x86/peephole_x86.cpp

```diff
@@ -0,0 +1,147 @@
+                         MachNode* (*new_root)(), uint inst0_rule, bool imm) {
+  MachNode* inst0 = block->get_node(block_index)->as_Mach();
+  assert(inst0->rule() == inst0_rule, "sanity");
+
+  OptoReg::Name dst = ra_->get_reg_first(inst0);
+  MachNode* inst1 = nullptr;
+  OptoReg::Name src1 = OptoReg::Bad;
+
+  if (inst0->in(1)->is_MachSpillCopy()) {
+    OptoReg::Name in = ra_->get_reg_first(inst0->in(1)->in(1));
+    if (OptoReg::is_reg(in) && OptoReg::as_VMReg(in)->is_Register()) {
+      inst1 = inst0->in(1)->as_Mach();
+      src1 = in;
+    }
+  }
+  if (inst1 == nullptr) {
+    return false;
+  }
+  assert(dst != src1, "");
+
+  // Only coalesce if inst1 is immediately followed by inst0
+  // Can be improved for more general cases
+  if (block_index < 1 || block->get_node(block_index - 1) != inst1) {
+    return false;
+  }
+  int inst1_index = block_index - 1;
+  Node* inst2;
+  if (imm) {
+    inst2 = nullptr;
+  } else {
+    inst2 = inst0->in(2);
+    if (inst2 == inst1) {
+      inst2 = inst2->in(1);
+    }
+  }
+
+  // See VM_Version::supports_fast_3op_lea()
+  if (!imm) {
+    Register rsrc1 = OptoReg::as_VMReg(src1)->as_Register();
+    Register rsrc2 = OptoReg::as_VMReg(ra_->get_reg_first(inst2))->as_Register();
+    if ((rsrc1 == rbp || rsrc1 == r13) && (rsrc2 == rbp || rsrc2 == r13)) {
+      return false;
+    }
+  }
+
+  // Go down the block to find the output proj node (the flag output) of inst0
+  int proj_index = -1;
+  Node* proj = nullptr;
+  for (uint pos = block_index + 1; pos < block->number_of_nodes(); pos++) {
+    Node* curr = block->get_node(pos);
+    if (curr->is_MachProj() && curr->in(0) == inst0) {
+      proj_index = pos;
+      proj = curr;
+      break;
+    }
+  }
+  assert(proj != nullptr, "");
+  // If some node uses the flag, cannot remove
+  if (proj->outcnt() > 0) {
+    return false;
+  }
+
+  MachNode* root = new_root();
+  // Assign register for the newly allocated node
+  ra_->set_oop(root, ra_->is_oop(inst0));
+  ra_->set_pair(root->_idx, ra_->get_reg_second(inst0), ra_->get_reg_first(inst0));
+
+  // Set input and output for the node
+  root->add_req(inst0->in(0));
+  root->add_req(inst1->in(1));
+  // No input for constant after matching
+  if (!imm) {
+    root->add_req(inst2);
+  }
+  inst0->replace_by(root);
+  proj->set_req(0, inst0);
+
+  // Initialize the operand array
+  root->_opnds[0] = inst0->_opnds[0]->clone();
+  root->_opnds[1] = inst0->_opnds[1]->clone();
+  root->_opnds[2] = inst0->_opnds[2]->clone();
+
+  // Modify the block
+  inst0->set_removed();
+  inst1->set_removed();
+  block->remove_node(proj_index);
+  block->remove_node(block_index);
+  block->remove_node(inst1_index);
+  block->insert_node(root, block_index - 1);
+
+  // Modify the CFG
+  cfg_->map_node_to_block(inst0, nullptr);
+  cfg_->map_node_to_block(inst1, nullptr);
+  cfg_->map_node_to_block(proj, nullptr);
+  cfg_->map_node_to_block(root, block);
+
```

Comment:
While it seems to be a little boiler-plate, I think a general helper that removes, inserts an arbitrary number of nodes, and connects the graph correctly is hard to write. What do you think? Thanks a lot.

Reply:
How does it work when there is no peepprocedure?  I would expect that all the boiler-plate details are taken care of by peepreplace.

Reply:
The code produced by `peepreplace` is really limited, it cannot be applied in this case because both the `MachSpillCopy` and `MachProj` are not match rules. Also, I think a generated code is more powerful than a helper function, so achieving something general with a helper function would be hard. I think we can refactor it into a helper if there are similarities arise later instead. What do you think?

Reply:
OK, agreed.